packages/compat/src/transaction.ts

```diff
@@ -0,0 +1,158 @@
+    }
+    const isSigner = message.isAccountSigner(accountIndex);
+    const isWritable = message.isAccountWritable(accountIndex);
+
+    const role = isSigner
+        ? isWritable
+            ? AccountRole.WRITABLE_SIGNER
+            : AccountRole.READONLY_SIGNER
+        : isWritable
+        ? AccountRole.WRITABLE
+        : AccountRole.READONLY;
+
+    return {
+        address: accountPublicKey.toBase58() as Base58EncodedAddress,
+        role,
+    };
+}
+
+// While the type `MessageCompiledInstruction` says data is a `Uint8Array`,
+// it is actually a Buffer in tests. Presumably this is a bug in the old web3js compile
+// message code. We will just assume it can be either here
+type MessageCompiledInstructionButDataMightBeABuffer = Omit<MessageCompiledInstruction, 'data'> & {
+    data?: Uint8Array | Buffer;
+};
+
+function convertInstruction(
+    message: VersionedMessage,
+    accountKeys: MessageAccountKeys,
+    instruction: MessageCompiledInstructionButDataMightBeABuffer
+): IInstruction {
+    const programAddressPublicKey = accountKeys.get(instruction.programIdIndex);
+    if (!programAddressPublicKey) {
+        // TODO coded error
+        throw new Error(`Could not find program address at index ${instruction.programIdIndex}`);
+    }
+
+    const accounts = instruction.accountKeyIndexes.map(accountIndex =>
+        convertAccount(message, accountKeys, accountIndex)
+    );
+
+    let data: Uint8Array | undefined;
+    const instructionData: Uint8Array | Buffer | undefined = instruction.data;
+    if (instructionData && instructionData.length > 0) {
+        if (instructionData instanceof Uint8Array) {
+            data = instructionData;
+        }
+        if (instructionData instanceof Buffer) {
+            data = new Uint8Array(instructionData.length);
+            instructionData.forEach((val, i) => {
+                data![i] = val;
+            });
+        }
+    }
+
+    return {
+        programAddress: programAddressPublicKey.toBase58() as Base58EncodedAddress,
+        ...(accounts.length ? { accounts } : {}),
+        ...(data ? { data } : {}),
+    };
+}
+
+function convertSignatures(
+    transaction: VersionedTransaction,
+    staticAccountKeys: PublicKey[]
+): ITransactionWithSignatures['signatures'] {
+    const signatures: Record<Base58EncodedAddress, Ed25519Signature> = {};
+
+    for (let i = 0; i < transaction.signatures.length; i++) {
+        const address = staticAccountKeys[i].toBase58() as Base58EncodedAddress;
+
+        // old web3js includes a fake all 0 signature if it hasn't been signed
+        // we don't do that for the new tx model. So just skip if it's all 0s
+        const allZeros = transaction.signatures[i].every(byte => byte === 0);
+        if (allZeros) continue;
+
+        const signature = transaction.signatures[i] as Ed25519Signature;
+        signatures[address] = signature;
+    }
+
+    return signatures;
+}
+
+export function fromOldVersionedTransactionWithBlockhash(
```

Comment:
I know you're probably using `Old` here because the transaction versions also use the term "legacy", but I gotta say I am much more in favor of "legacy" than "old" lol.

How can we rename this and still use some flavor of the terminology "Legacy Web3 JS"?

Reply:
Yea same, I prefer legacy to old too. But I think legacy in the context of transactions is just going to be way confusing. Especially because when address lookup tables get added here, this probably becomes 2 different functions (either explicitly or under the hood) depending if it is a legacy transaction version or v0.

Reply:
I wonder what it might look like if we just removed the word `Old` completely?

Or maybe we used `Deprecated` (idk if it really is yet). Or something worse: `LegacyWeb3JsVersionedTransaction1 lol.

Reply:
I'll drop Old, we can just have `fromVersionedTransactionWithBlockhash` etc. for now. Will have to figure something else out if we end up writing compatibility for the legacy-legacy `Transaction` class, but can figure that out if it happens!